### PR TITLE
fix: claimed in staking info endpoint

### PR DESCRIPTION
**Pull request by Imod7**
### Description
Closes https://github.com/paritytech/substrate-api-sidecar/issues/1433

This PR restores the field `claimed` (under `staking`) in the `staking-info` endpoint's response which was "broken". The root cause of this is explained [here](https://github.com/paritytech/substrate-api-sidecar/issues/1433#issuecomment-2075914389).

### For the Reviewer
To review this PR please refer to : 
1. the [hackMD](https://hackmd.io/@LwMsxe3-SFmNXxugAXOKgg/ryPwFoezyl): "_Staking PR - claimed field - Guide_".
1. the [README](https://github.com/paritytech/substrate-api-sidecar/pull/1445/files): "_STAKING_IMPLEMENTATION_DETAILS.md_". 

### Todos
- [x] Implement new logic based on [comment](https://github.com/paritytech/substrate-api-sidecar/pull/1445#issuecomment-2166042233)
- [x] Add a test that includes the era when the migration to the new logic happened
- [x] Add a test that includes also `partially claimed` eras & `erasStakersOverview` = null
- [x] Should return a total amount of eras (and their status) that are equal or less than `depth`. 
- [x] Merge master branch & update all historical tests (with new response type)
- [x] Handle the case in early blocks/eras where there is `lastReward` instead of `claimedRewards` under `stakingLedger`
- [x] Add the `nominator` logic mentioned in this [comment](https://github.com/paritytech/substrate-api-sidecar/pull/1445#issuecomment-2191412828)
- [x] Update docs
- [x] Add `IMPLEMENTATION_DETAILS` guide
- [x] Add reviewer's HackMD

### Changelog
This PR introduces **breaking changes** to the `staking-info` endpoint response by : 
1. Removing the returned field `lastReward` for early eras and adding `claimedRewards` updated accordingly.
2. Restoring the `claimedRewards` field name after the breaking change of renaming it to `legacyClaimedRewards`
3. Changing the structure/type of the returned `claimedRewards` field (under `staking`) : before we had an array of eras that were claimed, now we have an object with the list of eras and their corresponding status
4. Refactoring the whole logic that returns the `claimedRewards`
5. Adding the possibility to query nominator's claimed Rewards apart the validator's.

These changes should be explicitly mentioned in the corresponding Changelog and Release Notes of the release in which this PR is included.

### CODE REVIEW

**Comment by IkerAlus** [comment]
The change is reasonable. Sidecar user does not care of internal renaming changes since the retrieved info has the same meaning. 

Is this PR ready for review or are we waiting for the Todos list to be checked?

**Comment by Imod7** [comment]
> The change is reasonable. Sidecar user does not care of internal renaming changes since the retrieved info has the same meaning.
> 
> Is this PR ready for review or are we waiting for the Todos list to be checked?

Yes, ideally the todos needs to be completed before a review. However, if any review will speed things up then please feel free to add them.

**Comment by Imod7** [comment]
**Update**
After getting feedback from @Ank4n, for the `2nd check` mentioned in the `logic` section above, I concluded that the logic and thus the implementation in this PR should change to the following: 
- If `staking.erasStakersOverview.pageCount` == `query.staking.claimedRewards` -> era `claimed`
- If `staking.erasStakersOverview.pageCount > 0` &&  `query.staking.claimedRewards. length > 0` && `staking.erasStakersOverview.pageCount` != `query.staking.claimedRewards. length` -> era `partially claimed`
- If `overview == null` && `erasStakers > 0` -> `pageCount` = 1 : so then it depends on `query.staking.claimedRewards` value to see if era `claimed` or `unclaimed`

The response will also needs to be changed from a `Vec` of eras to a `Vec` of objects of type {era: eraNumber, status : claimed | unclaimed | partially claimed}

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 1]
```diff
@@ -66,6 +67,64 @@ export class AccountsStakingInfoService extends AbstractService {
 			);
 		}
 
+		let claimedRewardsEras: u32[] = [];
+		let claimedRewards: IEraStatus[] = [];
+
+		if ((stakingLedger as unknown as IStakingLedger)?.lastReward) {
+			const e = (stakingLedger as unknown as IStakingLedger)?.lastReward?.unwrap().toNumber();
```
Comment (marshacb) [inline src/services/accounts/AccountsStakingInfoService.ts:74]: Is it possible to unwrap a `None` value here?

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 2]
```diff
@@ -66,6 +67,64 @@ export class AccountsStakingInfoService extends AbstractService {
...
+		}
+		if (claimedRewardsEras) {
+			claimedRewards = claimedRewardsEras.map((element) => ({
+				era: element.toNumber(),
+				status: 'claimed',
+			}));
+		}
+		if (historicApi.query.staking?.claimedRewards) {
+			const currentEraMaybeOption = await historicApi.query.staking.currentEra();
+			const currentEra = currentEraMaybeOption.unwrap().toNumber();
```
Comment (marshacb) [inline src/services/accounts/AccountsStakingInfoService.ts:93]: same question as above

**Review by marshacb — APPROVED** [review]
LGTM! Just had a question about unwraps

File: src/services/accounts/AccountsStakingInfoService.ts (hunk 1 above)
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Very good point! I added a check in this [commit](https://github.com/paritytech/substrate-api-sidecar/pull/1445/commits/753f3eefde99a1fa750a972ba395ee53017131e5)

File: src/services/accounts/AccountsStakingInfoService.ts (hunk 2 above)
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:99]: You are right & thank you! In the same [commit](https://github.com/paritytech/substrate-api-sidecar/pull/1445/commits/753f3eefde99a1fa750a972ba395ee53017131e5) I added a check for `isNone` before unwrapping.

File: docs/src/openapi-v1.yaml [hunk 3]
```diff
@@ -4259,11 +4259,26 @@ components:
           format: unsignedInteger
         claimedRewards:
           type: array
-          description: Array of eras for which the stakers behind a validator have
-            claimed rewards. Only updated for _validators._
+          description: Array of objects, each containing an `era` and its corresponding `status`,
+            which represents the rewards status of the stakers backing a validator. Only updated
+            for _validators._ This array is populated with values from `stakingLedger.legacyClaimedRewards`
+            or `stakingLedger.claimedRewards`, as well as the `query.staking.claimedRewards` call, depending
+            on whether the queried block is before or after the migration. For more details on the implementation
```
Comment (IkerAlus) [inline docs/src/openapi-v1.yaml:4266]: Can we add a link here to the relevant polkadot-sdk release which included this migration?

File: docs/src/openapi-v1.yaml [hunk 4]
```diff
@@ -4259,11 +4259,26 @@ components:
...
+            and the migration, refer to the related PR and linked issue.
```
Comment (IkerAlus) [inline docs/src/openapi-v1.yaml:4267]: Where can the user find this info here?

**Comment by IkerAlus** [comment]
> The response time of the endpoint is significantly slower now

Do we have specific data for this statement? How long does it take to return the staking info in the any worst case scenario (for example, 84 eras in different pages and/or with different logic)?

**Comment by Imod7** [comment]
> > The response time of the endpoint is significantly slower now
> 
> Do we have specific data for this statement? How long does it take to return the staking info in the any worst case scenario (for example, 84 eras in different pages and/or with different logic)?

You are right, I didn't have any formal data when I wrote that. It was based on my observations from the tests I did and also on my knowledge of the calculations of each implementation. 
After adding some simple code to measure the execution time of `fetchAccountStakingInfo`, it looks like there are cases with similar runtimes (old code vs new code) and others have significant differences. However, these results can be influenced by factors such as the connection and the machine used. I am including below some results but to back this statement I need to run proper benchmarks. 

- `http://127.0.0.1:8080/accounts/GpyTMuLmG3ADWRxhZpHQh5rqMgNpFoNUyxA1DJAXfvsQ2Ly/staking-info?at=16869643` - kusama
    - old code : `608` ms
    - new code : `1.612` ms

- `http://127.0.0.1:8080/accounts/CmjHFdR59QAZMuyjDF5Sn4mwTgGbKMH2cErUFuf6UT51UwS/staking-info?at=22869643` - kusama
    - old code :  `372` ms
    - new code : `5.312` ms [this is an edge case- goes super slow]

- `http://127.0.0.1:8080/accounts/11VR4pF6c7kfBhfmuwwjWY3FodeYBKWx7ix2rsRCU2q6hqJ/staking-info?at=18157800` - polkadot
    - old code : `433` ms
    - new code : `408` ms

File: docs/src/openapi-v1.yaml (hunk 4 above)
Comment (Imod7) [reply docs/src/openapi-v1.yaml]: Added specific links of the PR and linked issue. Let me know if this covers your question. I also updated the schema description since I no longer return `lastReward`. Changes included in this [commit](https://github.com/paritytech/substrate-api-sidecar/pull/1445/commits/171e10a45749cfd1c73fddbe297efc8599d77ecb).

**Comment by Imod7** [comment]
@IkerAlus After our last convo on whether the status `partially claimed` is valid or not and some more feedback from the staking master @Ank4n : 

> _For a validator to get all commission, they should claim reward for all pages of an era._

Here is the next round of conclusions about `claimed`:

- When a user queries the `staking-info` endpoint with a **validator** address and at a specific **era**&block
   - The possible status is still : claimed, partially claimed or unclaimed 
- When a user queries a **nominator** address and at a specific **era**&block 
   - @IkerAlus you were right, the possible status can only be : claimed or unclaimed 

Currently, I am also considering that in both scenarios, it is useful to add the `undefined` status which is the case when we only have `lastReward` available (early eras). In such cases, we cannot determine if the queried era was claimed unless it matches the era noted in `lastReward`. If the queried era does not correspond to the `lastReward` era then claimed should be set as `undefined`.

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 5]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
 			throw new InternalServerError(
 				`Staking ledger could not be found for controller address "${controller.toString()}"`,
 			);
 		}
 
-		const numSlashingSpans = slashingSpansOption.isSome ? slashingSpansOption.unwrap().prior.length + 1 : 0;
+		// Fetching the list of validators that a nominator is nominating. This is only relevant for nominators.
+		// The stash account that we use as key is the nominator's stash account.
+		// https://polkadot.js.org/docs/substrate/storage/#nominatorsaccountid32-optionpalletstakingnominations
+		const nominations = historicApi.query.staking.nominators
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:85]: Nice one!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 6]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+	private async fetchErasStatusForValidator(
+		historicApi: ApiDecoration<'promise'>,
+		e: number,
+		stash: string,
+		claimedRewardsPerEra: Vec<u32>,
+		claimedRewards: IEraStatus<ValidatorStatus>[],
+	): Promise<IEraStatus<ValidatorStatus>[]> {
+		const erasStakersOverview: Option<SpStakingPagedExposureMetadata> =
+			await historicApi.query.staking.erasStakersOverview(e, stash);
+		let erasStakers: SpStakingExposure | null = null;
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Updated it and added `erasStakersOverview` as below.

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 7]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+				const currentEraMaybeOption = await historicApi.query.staking.currentEra();
+				if (currentEraMaybeOption.isNone) {
+					throw new InternalServerError('CurrentEra is None when Some was expected');
+				}
+				// Populating `claimedRewardsPerEra` for validator
+				const claimedRewardsPerEra: Vec<u32> = await historicApi.query.staking.claimedRewards(e, validatorStash);
+
+				// Doing similar checks as in fetchErasStatusForValidator function
+				// but with slight changes to adjust to nominator's case
+				const erasStakersOverview: Option<SpStakingPagedExposureMetadata> =
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Updated it and adjusted it to keep the type definitions.

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 8]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+				depth,
+				e,
+			);
+			claimedRewardsNom = claimedRewards as IEraStatus<NominatorStatus>[];
+			/**
+			 *  If the old calls are checked (which means `oldCallChecked` flag is true) and the new call
+			 * `query.staking.claimedRewards` is available then we go into this check.
+			 */
+			if (historicApi.query.staking?.claimedRewards && oldCallChecked) {
+				const currentEraMaybeOption = await historicApi.query.staking.currentEra();
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Updated it! Thank you!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 9]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+	}
+
+	private async fetchErasStatusForNominator(
+		historicApi: ApiDecoration<'promise'>,
+		e: number,
+		depth: number,
+		eraStart: number,
+		claimedRewardsNom: IEraStatus<NominatorStatus>[],
+		validatorsUnwrapped: string[],
+		nominatorStash: string,
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:243]: Updated this!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 10]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+					const validatorsUnwrapped = nominatingValidators.unwrap().targets.toHuman() as string[];
+
+					const [era, claimedRewardsNom1] = await this.fetchErasStatusForNominator(
+						historicApi,
+						e,
+						depth,
+						eraStart,
+						claimedRewardsNom,
+						validatorsUnwrapped,
+						stash,
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:151]: Done!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 11]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+				oldCallChecked = oldCallCheck;
+				e = era;
+			} else {
+				oldCallChecked = true;
+			}
+
+			// Checking if the new call is available then I can check if rewards of nominator are claimed or not.
+			// If not available, I will set the status to 'undefined'.
+			if (historicApi.query.staking?.claimedRewards && oldCallChecked) {
+				const currentEraOption = await historicApi.query.staking.currentEra();
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Good one! Done!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 12]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+						historicApi,
+						e,
+						stash,
+						claimedRewardsPerEra,
+						claimedRewards,
+					);
+				} else {
+					// To verify the reward status `claimed` of an era for a Nominator's account,
+					// we need to check the status of that era in one of their associated Validators' accounts.
+					const nominatingValidators = await historicApi.query.staking.nominators(stash);
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Another good one! Done!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 13]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+			}
+		}
+		return 'undefined';
+	}
+
+	/**
+	 * This function determines which eras to check for claimed rewards
+	 * by returning the starting era (`eraStart`) and the number of eras to check (`depth`)
+	 */
+	private async fetchErasInfo(api: ApiDecoration<'promise'>, historicApi: ApiDecoration<'promise'>) {
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Yes, nice! Done!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 14]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+		// https://polkadot.js.org/docs/substrate/storage/#nominatorsaccountid32-optionpalletstakingnominations
+		const nominations = historicApi.query.staking.nominators
+			? (await historicApi.query.staking.nominators(stash)).unwrapOr(null)
+			: null;
+
+		// Initializing two arrays to store the status of claimed rewards per era for validators and for nominators.
+		let claimedRewards: IEraStatus<ValidatorStatus | NominatorStatus>[] = [];
+		let claimedRewardsNom: IEraStatus<NominatorStatus>[] = [];
+
+		const [eraStart, depth] = await this.fetchErasInfo(api, historicApi);
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Updated it!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 15]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+		// The stash account that we use as key is the nominator's stash account.
+		// https://polkadot.js.org/docs/substrate/storage/#nominatorsaccountid32-optionpalletstakingnominations
+		const nominations = historicApi.query.staking.nominators
+			? (await historicApi.query.staking.nominators(stash)).unwrapOr(null)
+			: null;
+
+		// Initializing two arrays to store the status of claimed rewards per era for validators and for nominators.
+		let claimedRewards: IEraStatus<ValidatorStatus | NominatorStatus>[] = [];
+		let claimedRewardsNom: IEraStatus<NominatorStatus>[] = [];
+
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:95]: We discussed this and we agreed on keeping it as it is with `fetchErasStatusForNominator` and `fetchErasStatusForValidator`. If you find other major implications in performance or readability, I am open to discuss this again. Resolving for now.

[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:95]: I think this is the same as mentioned above right? Resolving for now with the same [reason](https://github.com/paritytech/substrate-api-sidecar/pull/1445#discussion_r1914813709).

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 16]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+				claimedRewardsNom = claimedRewardsOld;
+				oldCallChecked = oldCallCheck;
+				e = era;
+			} else {
+				oldCallChecked = true;
+			}
+
+			// Checking if the new call is available then I can check if rewards of nominator are claimed or not.
+			// If not available, I will set the status to 'undefined'.
+			if (historicApi.query.staking?.claimedRewards && oldCallChecked) {
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts:269]: Yes, I know this is not ideal but it was done on purpose. It was my only way to check for the new call. If it does not exist then I do not want to go into that statement at all because all the code there is for the new call. But if you find a better way then I am definitely on board!

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 17]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+						stash,
+					);
+					e = era;
+					claimedRewardsNom = claimedRewardsNom1;
+				}
+			} else {
+				break;
+			}
 		}
+		if (!isValidator) {
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Removed it.

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 18]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+			 *  If the old calls are checked (which means `oldCallChecked` flag is true) and the new call
+			 * `query.staking.claimedRewards` is available then we go into this check.
+			 */
+			if (historicApi.query.staking?.claimedRewards && oldCallChecked) {
+				if (currentEraOption.isNone) {
+					throw new InternalServerError('CurrentEra is None when Some was expected');
+				}
+
+				// Populating `claimedRewardsPerEra` for validator
+				const claimedRewardsPerEra: Vec<u32> = await historicApi.query.staking.claimedRewards(e, stash);
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Updated it

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 19]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+
+			// Checking if the new call is available then I can check if rewards of nominator are claimed or not.
+			// If not available, I will set the status to 'undefined'.
+			if (historicApi.query.staking?.claimedRewards && oldCallChecked) {
+				const currentEraOption = await historicApi.query.staking.currentEra();
+				if (currentEraOption.isNone) {
+					throw new InternalServerError('CurrentEra is None when Some was expected');
+				}
+				// Populating `claimedRewardsPerEra` for validator
+				const claimedRewardsPerEra: Vec<u32> = await historicApi.query.staking.claimedRewards(e, validatorStash);
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Done

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 20]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+		historicApi: ApiDecoration<'promise'>,
+		e: number,
+		stash: string,
+		claimedRewardsPerEra: Vec<u32>,
+		claimedRewards: IEraStatus<ValidatorStatus>[],
+	): Promise<IEraStatus<ValidatorStatus>[]> {
+		const [erasStakersOverview, erasStakers]: [Option<SpStakingPagedExposureMetadata>, SpStakingExposure | null] =
+			await Promise.all([
+				historicApi.query.staking.erasStakersOverview(e, stash),
+				historicApi.query.staking?.erasStakers ? historicApi.query.staking.erasStakers(e, stash) : null,
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Done

File: src/services/accounts/AccountsStakingInfoService.ts [hunk 21]
```diff
@@ -48,39 +58,422 @@ export class AccountsStakingInfoService extends AbstractService {
...
+	}
+
+	private async fetchErasStatusForValidator(
+		historicApi: ApiDecoration<'promise'>,
+		e: number,
+		stash: string,
+		claimedRewardsPerEra: Vec<u32>,
+		claimedRewards: IEraStatus<ValidatorStatus>[],
+	): Promise<IEraStatus<ValidatorStatus>[]> {
+		const [erasStakersOverview, erasStakers]: [Option<SpStakingPagedExposureMetadata>, SpStakingExposure | null] =
```
[the remark this answers is not in the document]
Comment (Imod7) [reply src/services/accounts/AccountsStakingInfoService.ts]: Done

**Comment by Imod7** [comment]
Thank you @filvecchiato for the super thorough and detailed review. All your notes on the performance improvements were well noted and implemented. Your questions, doubts and thoughts on the logic were excellent and super helpful in improving the implementation guide for future readers. I greatly appreciate the time, effort and brain matter (due to its complexity) you put into this review. A big thank you!
